include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,377 @@
+                        return &(variable->second);
+                    }
+                }
+                return nullptr;
+            }
+
+            // replace or create the current value
+            void set(std::any value)
+            {
+                // release value, with the swapped value, outside of the lock
+                {
+                    auto lock = winrt::slim_lock_guard(s_lock);
+                    auto storage = s_apartmentStorage.find(test_hook::GetApartmentId());
+                    FAIL_FAST_IF(storage == s_apartmentStorage.end());
+                    auto& variables = storage->second;
+                    auto variable = variables.find(this);
+                    FAIL_FAST_IF(variable == variables.end());
+                    variable->second.swap(value);
+                }
+            }
+
+            // remove any current value
+            void clear()
+            {
+                auto lock = winrt::slim_lock_guard(s_lock);
+                auto storage = s_apartmentStorage.find(test_hook::GetApartmentId());
+                if (storage != s_apartmentStorage.end())
+                {
+                    auto& variables = storage->second;
+                    variables.erase(this);
+                }
+            }
+
+            static size_t apartment_count()
+            {
+                auto lock = winrt::slim_lock_guard(s_lock);
+                return s_apartmentStorage.size();
+            }
+
+            static size_t current_apartment_variable_count()
+            {
+                auto lock = winrt::slim_lock_guard(s_lock);
+                auto storage = s_apartmentStorage.find(test_hook::GetApartmentId());
+                if (storage != s_apartmentStorage.end())
+                {
+                    return storage->second.size();
+                }
+                return 0;
+            }
+        };
+    }
+
+    template<typename T, typename test_hook = apt_var_platform>
+    struct apartment_variable : details::apartment_variable_base<test_hook>
+    {
+        using base = details::apartment_variable_base<test_hook>;
+
+        // get current value or throw if no value has been set
+        T& get_existing() { return std::any_cast<T&>(base::get_existing()); }
+
+        // get current value or default-construct one on demand
+        T& get_or_create()
+        {
+            return get_or_create([]() { return T{}; });
+        }
+
+        // get current value or custom-construct one on demand
+        T& get_or_create(T(*creator)())
+        {
+            return std::any_cast<T&>(base::get_or_create([creator]()
+            {
+                return std::any(creator());
+            }));
+        }
+
+        // get pointer to current value or nullptr if no value has been set
+        T* get_if() { return std::any_cast<T>(base::get_if()); }
+
+        // replace or create the current value
+        void set(T&& value) { return base::set(std::forward<T>(value)); }

```

Comment:
Incorrect use of universal reference.

```suggestion
        template<typename U> void set(U&& value) { return base::set(std::forward<U>(value)); }

```

Reply:
"forwarding reference" ;-). Fixed.